servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -0,0 +1,353 @@
+                            DelayedTermination nextTerm = cDelayedTerm.copyUsing(collector);
+                            if (maybeCollectorUpdater.compareAndSet(this, ccMaybeCollector, nextTerm)) {
+                                boundariesSubscriber.tryTerminate(nextTerm);
+                                return;
+                            }
+                        }
+                    }
+                    break;
+                }
+            }
+
+            if (pending == 0) {
+                requestMore();
+            }
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            boundariesSubscriber.terminateFromItemsSource(TerminalNotification.error(t));
+        }
+
+        @Override
+        public void onComplete() {
+            boundariesSubscriber.terminateFromItemsSource(TerminalNotification.complete());
+        }
+
+        private void requestMore() {
+            pendingItemsUpdater.accumulateAndGet(this, bufferSizeHint, FlowControlUtils::addWithOverflowProtection);
+            tSubscription.request(bufferSizeHint);
+        }
+
+        private final class BoundariesSubscriber extends DelayedCancellable
+                implements Subscriber<Collector<T, B>> {
+            private final Subscriber<? super B> target;
+            private final Cancellable tCancellable;
+
+            BoundariesSubscriber(final Subscriber<? super B> target, final Cancellable tCancellable) {
+                this.target = target;
+                this.tCancellable = tCancellable;
+            }
+
+            @Override
+            public void onSubscribe(final Subscription subscription) {
+                ConcurrentSubscription bs = ConcurrentSubscription.wrap(new Subscription() {
+                    @Override
+                    public void request(final long n) {
+                        if (!isRequestNValid(n)) {
+                            // This code is deliberately not intending to avoid potentially concurrent calls to the
+                            // subscriber as the specification is already violated. This avoids complicating this
+                            // state machine for unexpected specification violations.
+                            Object oldUpdater = maybeCollectorUpdater.getAndSet(BufferSubscriber.this, TERMINATED);
+                            if (oldUpdater != TERMINATED) {
+                                subscription.cancel();
+                                safeOnError(target, newExceptionForInvalidRequestN(n));
+                            }
+                            return;
+                        }
+                        if (pendingBuffersUpdater.getAndAccumulate(BufferSubscriber.this, n,
+                                FlowControlUtils::addWithOverflowProtection) == 0) {
+                            Object cMaybeCollector = maybeCollector;
+                            if (cMaybeCollector instanceof DelayedTermination) {
```

Comment:
`onError` maybe delivered to `BufferSubscriber`, but it isn't clear what provides visibility here such that the event will be delivered.

```
T1 - BufferSubscriber#onSubscribe(..)
  toSource(boundaries).subscribe(boundariesSubscriber)

...
  BufferSubscriber#onError(..)
  boundariesSubscriber.terminateFromItemsSource
  terminate(this, terminalNotification)
  maybeCollectorUpdater.CAS(null, DelayedTermination)
  tryTerminate(..)
  cPendingBuffers = pendingBuffers -> 0
  return;
  
// Some later time...
T2 - BoundariesSubscriber#onSubscribe
  target.onSubscribe(bs)
  bs.request(1)
  pendingBuffersUpdater.GAA() -> 1
  cMaybeCollector = maybeCollector -> null // no-visibility?
  subscription.request(1)

...
  BoundariesSubscriber#onNext(..)
  cMaybeCollector = maybeCollector -> DelayedTermination
  return;
```

the unconditional dependency on `pendingBuffersUpdater` has a nice side effect the way the code is currently structured in that it also prevents `onError` from coming before `onSubscribe`, but has some other un-intended consequences.